Sources/NIOTestUtils/EventCounterHandler.swift

```diff
@@ -0,0 +1,363 @@
+    }
+
+    /// Returns the number of `connect` events seen so far in the `ChannelPipeline`.
+    public var connectCalls: Int {
+        return self._connectCalls.load()
+    }
+
+    /// Returns the number of `write` events seen so far in the `ChannelPipeline`.
+    public var writeCalls: Int {
+        return self._writeCalls.load()
+    }
+
+    /// Returns the number of `flush` events seen so far in the `ChannelPipeline`.
+    public var flushCalls: Int {
+        return self._flushCalls.load()
+    }
+
+    /// Returns the number of `read` events seen so far in the `ChannelPipeline`.
+    public var readCalls: Int {
+        return self._readCalls.load()
+    }
+
+    /// Returns the number of `close` events seen so far in the `ChannelPipeline`.
+    public var closeCalls: Int {
+        return self._closeCalls.load()
+    }
+
+    /// Returns the number of `triggerUserOutboundEvent` events seen so far in the `ChannelPipeline`.
+    public var triggerUserOutboundEventCalls: Int {
+        return self._triggerUserOutboundEventCalls.load()
+    }
+
+    /// Validate some basic assumptions about the number of events and if any of those assumptions are violated, throw
+    /// an error.
+    ///
+    /// For example, any `Channel` should only be registered with the `Selector` at most once. That means any
+    /// `ChannelHandler` should see no more than one `register` event. If `EventCounterHandler` sees more than one
+    /// `register` event and you call `checkValidity`, it will throw `EventCounterHandler.ValidityError` with an
+    /// appropriate explanation.
+    ///
+    /// - note: This API is thread-safe, you may call it from any thread. The results of this API may vary though if you
+    ///         call it whilst the `Channel` this `ChannelHandler` is in is still in use.
+    public func checkValidity() throws {
+        guard self.channelRegisteredCalls <= 1 else {
+            throw ValidityError("channelRegistered should be called no more than once")
+        }
+        guard self.channelUnregisteredCalls <= 1 else {
+            throw ValidityError("channelUnregistered should be called no more than once")
+        }
+        guard self.channelActiveCalls <= 1 else {
+            throw ValidityError("channelActive should be called no more than once")
+        }
+        guard self.channelInactiveCalls <= 1 else {
+            throw ValidityError("channelInactive should be called no more than once")
+        }
+        guard self.registerCalls <= 1 else {
+            throw ValidityError("register should be called no more than once")
+        }
+        guard self.bindCalls <= 1 else {
+            throw ValidityError("bind should be called no more than once")
+        }
+        guard self.connectCalls <= 1 else {
+            throw ValidityError("connect should be called no more than once")
+        }
+        guard self.closeCalls <= 1 else {
+            throw ValidityError("close should be called no more than once")
```

Comment:
This assertion is not universally correct: no bugs will occur if `close` is called more than once, and I don't know if I'd class it as the same kind of error as the others.

Reply:
@Lukasa thanks, that's right, will remove